<a href="https://colab.research.google.com/github/prisar/ai_notebooks/blob/main/nb_097.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Summarize the video located at "gs://mrc-quant-ml-video-analysis/videoplayback.mp4" using a temporal-aware summarization method involving video chunking, Whisper transcription, vector encoding, and a memory-augmented LLM, and evaluate the faithfulness, temporal coherence, and semantic coverage of the resulting summary.

## Setup environment

### Subtask:
Install necessary libraries (Whisper, LlamaIndex, etc.) and set up authentication for accessing the video from GCS.


**Reasoning**:
Install the necessary libraries using pip.



In [1]:
!pip install whisper llama-index google-cloud-storage moviepy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 kB 2.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 83.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 303.3/303.3 kB 30.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.0/41.0 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 313.2/313.2 kB 31.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 79.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.9/63.9 kB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.1/137.1 kB 15.2 MB/s eta 0:00:00
  Created wheel for whisper: filename=whisper-1.1.10-py3-none-any.whl size=41120 sha256=9ca8592c21e2f5e428874f70ce6efff63dec4f54fc2bf90ac049f66c8227ba27
  Stored in directory: /root/.cache/pip/wheels/21/65/ee/4e6672aabfa486d3341a39a04f8f87c77e51561

**Reasoning**:
Authenticate to Google Cloud Storage. Since this is running in a Google Cloud environment, the recommended way is to use the default credentials provided by the environment.



In [1]:
from google.cloud import storage

# Initialize a client using the default credentials
storage_client = storage.Client()

print("Google Cloud Storage authentication successful.")

Google Cloud Storage authentication successful.


## Download video

### Subtask:
Download the video from the specified GCS path to the Colab environment.


**Reasoning**:
Define the GCS bucket and blob names, the destination file name, and then download the blob to the specified destination file using the storage client.



In [2]:
# Define the GCS bucket name and source blob name
bucket_name = "mrc-quant-ml-video-analysis"
source_blob_name = "videoplayback.mp4"

# Define the destination file name in the Colab environment
destination_file_name = "downloaded_video.mp4"

# Get the bucket object
bucket = storage_client.get_bucket(bucket_name)

# Get the blob object (the video file) from the bucket
blob = bucket.blob(source_blob_name)

# Download the blob to the specified destination file
blob.download_to_filename(destination_file_name)

# Print a confirmation message
print(f"Video downloaded successfully to {destination_file_name}")

RefreshError: ("Failed to retrieve http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true from the Google Compute Engine metadata service. Status: 404 Response:\nb''", <google.auth.transport.requests._Response object at 0x7849e41d7590>)

## Download video retry

### Subtask:
Retry downloading the video from the specified GCS path to the Colab environment, addressing potential authentication issues.


**Reasoning**:
Attempt to download the video again, ensuring the storage client is initialized and handling potential authentication issues by letting the default credentials try again. If it fails, provide details about the error.



In [3]:
# Define the GCS bucket name and source blob name
bucket_name = "mrc-quant-ml-video-analysis"
source_blob_name = "videoplayback.mp4"

# Define the destination file name in the Colab environment
destination_file_name = "downloaded_video.mp4"

try:
    # Get the bucket object
    bucket = storage_client.get_bucket(bucket_name)

    # Get the blob object (the video file) from the bucket
    blob = bucket.blob(source_blob_name)

    # Download the blob to the specified destination file
    blob.download_to_filename(destination_file_name)

    # Print a confirmation message
    print(f"Video downloaded successfully to {destination_file_name}")

except Exception as e:
    print(f"Error downloading video: {e}")
    print("Please ensure your Colab environment is properly authenticated to access the GCS bucket.")


Error downloading video: ("Failed to retrieve http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/?recursive=true from the Google Compute Engine metadata service. Status: 404 Response:\nb''", <google.auth.transport.requests._Response object at 0x7849e4333a90>)
Please ensure your Colab environment is properly authenticated to access the GCS bucket.


## Summary:

### Data Analysis Key Findings

*   All required libraries for video processing and summarization (`whisper`, `llama-index`, `google-cloud-storage`, `moviepy`) were successfully installed.
*   An initial attempt to download the video from Google Cloud Storage (GCS) failed due to an authentication error (`RefreshError`), indicating a problem with credentials in the Colab environment.
*   A retry of the video download also failed, showing a persistent authentication issue (Status: 404 when trying to retrieve metadata), suggesting that the Colab environment's default credentials are not sufficient or properly configured to access the specified GCS bucket.

### Insights or Next Steps

*   The task of summarizing the video cannot proceed without successfully downloading the video.
*   The next step is to troubleshoot and resolve the Google Cloud Storage authentication issue in the Colab environment, potentially by explicitly authenticating with service account credentials that have access to the GCS bucket.
